packages/circuits/src/proofs/DrawHand.circom

```diff
@@ -0,0 +1,165 @@
+
+/**
+ * This circuit is responsible for proving that the player drew the correct initial hand of cards.
+ *
+ * Parameters:
+ * - `levels` is the base-2 log of the max size of the deck (6 for a deck of max 64 cards)
+ * - `initialHandSize` is the amount of cards to draw
+ *
+ * Public inputs:
+ * - `initialDeck` are the cards in the initial deck listing (padded with 255 to reach a size of 2**levels)
+ * - `deckRoot` is the Merkle root of the deck after drawing the cards
+ * - `handRoot` is the Merkleroot of the hand the player drew
+ * - `saltHash` is the hash of player's secret salt, acting as an on-chain commitment to that salt
+ * - `publicRandom` is a public randomness value available on-chain
+ *
+ * Private inputs:
+ * - `hand` are the cards drawn by the player
+ * - `deck` are the cards in the deck after removing the drawn card
+ *   (replacing them with cards from the back of the deck and shrinking the deck)
+ * - `salt` is the player's secret salt
+ *
+ * About 222k Plonk constraints.
+ */
+template DrawHand(levels, initialHandSize) {
+
+    // public inputs
+    signal input initialDeck[2**levels];
+    signal input deckRoot;
+    signal input handRoot;
+    signal input saltHash;
+    signal input publicRandom;
+
+    // private inputs
+    signal input salt;
+    signal input deck[2**levels];
+    signal input hand[2**levels];
+
+    // verify the private salt matches the public salt hash
+    component checkSalt = MiMCSponge(1, 220, 1);
+    checkSalt.ins[0] <== salt;
+    checkSalt.k <== 0;
+    saltHash === checkSalt.outs[0];
+
+    // compute randomness from private salt and public randomness (publicRandom)
+    component randomness = MiMCSponge(2, 220, 1);
+    randomness.ins[0] <== salt;
+    randomness.ins[1] <== publicRandom;
+    randomness.k <== 0;
+
+    component drawCards[initialHandSize];
+    signal divisors[initialHandSize];
+
+    // This will contain the deck as one card is being drawn (swapped out for last card) at a time.
+    signal intermediateDecks[initialHandSize+1][2**levels];
+    intermediateDecks[0] <== initialDeck;
+
+    for (var i = 0; i < initialHandSize; i++) {
+
+        // NOTE: This is buggy.
+        // This should use a new public signal giving the deck size, and be `deckSize - 1 - i`.
+        // Unfortunately, doing so would break the circuit because we rely on lastIndex being a
+        // compile-time constant. This is fixable, but requires to double the number of deck iterations
+        // (need to add an extra iteration to pick out the current last card).
+        var lastIndex = 2**levels - 1 - i;
+
+        drawCards[i] = RemoveIndex(levels, lastIndex + 1);
+
+        // pick out a random card — we need to do the dance to prove the modulus
+        drawCards[i].index <-- randomness.outs[0] % lastIndex;
+        divisors[i] <-- randomness.outs[0] / lastIndex;
+        randomness.outs[0] === divisors[i] * lastIndex + drawCards[i].index;
```

Comment:
So apparently we will need to use `randomness.outs[0] \ lastIndex;` to get the quotient. The operator `/` will be modulo by the prime field, which makes the constraint incorrect. 

Check out the difference between `\` and `/` here: https://docs.circom.io/circom-language/basic-operators/#arithmetic-operators